### PR TITLE
Add ctrl+p editor prompt (quick actions)

### PR DESCRIPTION
**Pull request by ChillerDragon**
closes https://github.com/ddnet/ddnet/issues/7933
follow up to https://github.com/ddnet/ddnet/pull/8358

I only added a few actions for now because I don't think the prompt and all the actions belong together in one pr. I will contribute more actions as soon as the prompt it self is merged.

image

https://github.com/user-attachments/assets/03def380-34dc-40fb-985c-91e9d7789a3b

<details>
<summary>Outdated video with old look</summary>

https://github.com/user-attachments/assets/dc4e184f-d955-4e8a-9d5e-0d350155644e

</details>

## Checklist

- [x] Tested the change ingame
- [x] Provided screenshots if it is a visual change
- [x] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [ ] Considered possible null pointers and out of bounds array indexing
- [ ] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -2360,8 +2360,13 @@ if(CLIENT)
     mapitems/sound.cpp
     mapitems/sound.h
     popups.cpp
+    prompt.cpp
+    prompt.h
     proof_mode.cpp
     proof_mode.h
+    quick_action.h
+    quick_actions.cpp
+    quick_actions.h
```
Comment (BlaiZephyr) [inline CMakeLists.txt:2369]: Multiple include

**Review by BlaiZephyr — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 1 above)
Comment (ChillerDragon) [reply CMakeLists.txt:2369]: s

Comment (ChillerDragon) [reply CMakeLists.txt:2369]: One header is the class that describes a quick action. And the other is the list of all registered quick actions.

**Comment by jxsl13** [comment]
can you remove the rounded corners at the top ._.

File: CMakeLists.txt (hunk 1 above)
Comment (BlaiZephyr) [reply CMakeLists.txt:2369]: why not quick_action_list.h then :P - i was tired and didn't even see the s. gg

Comment (ChillerDragon) [reply CMakeLists.txt:2369]: Yea I agree the naming is not perfect. Not sure if adding "list" is any better tho.

File: src/game/editor/prompt.cpp [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+bool CPrompt::OnInput(const IInput::CEvent &Event)
+{
+	if(Input()->ModifierIsPressed() && Input()->KeyIsPressed(KEY_P))
+	{
+		SetActive();
+	}
+	return false;
+}
+
+void CPrompt::OnRender(CUIRect View)
```
Comment (Robyt3) [inline src/game/editor/prompt.cpp:51]: Regarding the layout, screenshot for reference:

screenshot_2024-07-25_21-54-21

[As stated before](https://github.com/ddnet/ddnet/pull/8358#discussion_r1672692865), it shouldn't be rendered over the map view. Sorry, I didn't see your response to my comment before.

> Whats wrong with rendering over the map view? Thats seems to be the intended way to use the component system.

The issue with rendering over the map view is that it looks weird and depending on the configuration makes it very hard to use the prompt feature:

screenshot_2024-07-25_21-58-00

It also looks better like a regular modal dialog with the darker background overlayed over the rest of the UI.

The map settings error dialog is also rendered like a dialog in `CEditor::RenderMapSettingsErrorDialog` not with the component `OnRender` function, so I think this is the intended way to render modal dialogs that don't overlay each other.

The text input should have the same height as the other editor buttons in the menu bars for consistency, it doesn't need to be so large.

The description text is currently hard to read due to the low contrast, both for the selected and unselected entries.

The margins on the left and right sides should also be consistent.

**Review by Robyt3 — CHANGES_REQUESTED** [review]

**Comment by ChillerDragon** [comment]
Fixed the prompt overlay. Now looks like this:

image

image

**Comment by ChillerDragon** [comment]
Fuzzy match is not perfect yet. It only filters it does not sort yet. There should be some sense of score.

image


Searching "roo" should for sure put "Proof" on the first hit.

**Comment by ChillerDragon** [comment]
There is still no sense of score but that should be fine I think. I fixed the specific "Proof" case yielding too many results because duplicated letters were not strictly checked. Now `roo` only finds `Proof`

image

**Comment by ChillerDragon** [comment]
I added the first action that can be in state disabled. One can now select a mapres for a tile layer without using the mouse.
But the option will only be listed if a tile layer is selected (no game layer).


https://github.com/user-attachments/assets/c5698231-2117-4e38-8458-dc3675518856

**Comment by ChillerDragon** [comment]
Solved conflicts with master. Still waiting for a review (since last month) and ready to be merged from my side.

**Review by Robyt3 — APPROVED** [review]

**Review by def- — APPROVED** [review]

**Pull request merged**